### PR TITLE
Trezor Unlock + Deterministic Wallet Groundwork

### PR DESCRIPTION
**Pull request by wbobeirne**
## What This Does

This is the first pass at hardware wallet support, starting with Trezor, since that's the one I own. It allows for the unlocking of, and signing transactions for Trezor wallets. It also lays down a lot of the groundwork for future deterministic path generated wallets (Mnemonic, ledger, etc.)

Don't let the line count fool you, this PR isn't _that_ big. The bulk of it was because of `common/vendor/trezor-connect.js`. It's not on NPM, so it has to be loaded into the project directly. Currently in talks with them about the future of Trezor Connect.

The wallet modal was modeled after #129, not v3 directly, so there are some noteworthy differences.

## What This Doesn't Do

* Other hardware wallets
* Nearly enough unit testing, would love to figure out a way to do this better
* Message signing
    * There's no base method for this on BaseWallet, and I wasn't sure how solid the private key implementation was, so I figured I'd hold off on this until it came up
* Guaranteed accurate transaction signing
    * I implemented it as best I could from v3, but without broadcasting, it's hard to be sure I'm doing it right. Would love some help in auditing this.

## Basic Technical Overview

This is how Trezor does everything, and presumably the other HW wallets will have a similar flow:

* All interfacing with Trezor happens at the component layer in `components/WalletDecrypt/Trezor.jsx` via TrezorConnect
* Callbacks from TrezorConnect update props sent to `components/WalletDecrypt/DeterministicWalletsModal.jsx`
* The modal component is connected, and fires off actions to the `deterministicWallets` reducer / saga
* Changes that happen in the modal (changing path, selecting address) are bubbled back up to `Trezor.jsx`
* Once a wallet is selected, `Trezor.jsx` calls `onUnlock` and creates a new `TrezorWallet` instance
* `TrezorWallet` extends the `DeterministicWallet`, and only differs for `signRawTransaction` (And presumably `signMessage` in the future)

## Trezor in Action

_GIFs were recorded at reduced sizes, and addresses / amounts blurred, so don't worry if things look a little odd_

### Connecting

trezor-connect

---

### Token Values

trezor-tokens

---

### Paging to More Addresses

trezor-paging

---

### Path Selection

trezor-path

---

### Unlocking

trezor-unlocking

### CODE REVIEW

File: common/components/WalletDecrypt/Trezor.jsx [hunk 1]
```diff
@@ -1,26 +1,139 @@
+// @flow
+import './Trezor.scss';
 import React, { Component } from 'react';
 import translate from 'translations';
+import TrezorConnect from 'vendor/trezor-connect';
+import DeterministicWalletsModal from './DeterministicWalletsModal';
+import TrezorWallet from 'libs/wallet/trezor';
+
+/* eslint-disable quotes */
+const TREZOR_PATHS = [
```
Comment (dternyak) [inline common/components/WalletDecrypt/Trezor.jsx]: consider moving out of component (and into config)

File: common/libs/transaction.js [hunk 2]
```diff
@@ -47,8 +48,8 @@ export function getTransactionFields(tx: EthTx) {
   const [nonce, gasPrice, gasLimit, to, value, data, v, r, s] = tx.toJSON();
 
   return {
-    // No value comes back as '0x', but most things expect '0x0'
-    value: value === '0x' ? '0x0' : value,
+    // No value comes back as '0x', but most things expect '0x00'
```
Comment (dternyak) [inline common/libs/transaction.js]: any link to this? seems like we're changing it once already, so would be good to know where this info is coming from.

File: common/libs/transaction.js [hunk 3]
```diff
@@ -121,13 +122,14 @@ export async function generateTransaction(
 
   // Generate the raw transaction
   const txCount = await node.getTransactionCount(tx.from);
+  const cleanHex = hex => addHexPrefix(padToEven(stripHex(hex)));
```
Comment (dternyak) [inline common/libs/transaction.js]: What kind of padding are we doing here? Maybe some comments to explain?

File: common/libs/validators.js [hunk 4]
```diff
@@ -139,3 +139,7 @@ export function isValidRawTx(rawTx: RawTransaction): boolean {
 
   return true;
 }
+
+export function isValidPath(dPath: string) {
+  return dPath.split('\'/').length === 4;
```
Comment (dternyak) [inline common/libs/validators.js]: maybe a comment here explaining what's going on. I get that the deterministic path is being split, but its not extremely clear why 4 is important

File: common/libs/transaction.js (hunk 2 above)
Comment (wbobeirne) [reply common/libs/transaction.js]: Yeah, would definitely be good to figure out the "correct" behavior here. I just used these values based on complaining libraries. `0x0` from `0x` came from the node, `0x00` came from Trezor wanting padded values.

File: common/components/WalletDecrypt/Trezor.jsx (hunk 1 above)
Comment (wbobeirne) [reply common/components/WalletDecrypt/Trezor.jsx]: 👍

File: common/libs/transaction.js (hunk 3 above)
Comment (wbobeirne) [reply common/libs/transaction.js]: This was taken from v3's `sanitizeHex` behavior, Trezor wanted %2 data. I can definitely add some comments to v3, but I don't truly understand the reasoning.

File: common/libs/validators.js (hunk 4 above)
Comment (wbobeirne) [reply common/libs/validators.js]: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki has more information, and test cases, I'll try to add a useful comment and implement some of its test cases.

**Comment by wbobeirne** [comment]
Added clarifying comments, didn't end up implementing the bip32 test cases because they were more about `hdk.derive` and `publicToAddress` from `hdkey` and `ethereumjs-util` respectively.

Ready to have this merged whenever!

**Comment by jamesob** [comment]
:tada:

**Pull request merged**